# Assignment - Convolutional Neural Networks

Develop a CNN model using the CIFAR-100 dataset which has 32x32 images from 100 different classes. Try to increase the accuracy of the model to over 0.95. To do this, you can change the number of epochs, the number of convolution and dense layers, the kernel size, the number of filters, the padding and stride values.

You can download the dataset with the code from keras.datasets import cifar100.

In [1]:
from keras.models import Sequential
from tensorflow.keras.datasets import cifar100
from keras.layers import Dense, Dropout, Conv2D, MaxPool2D, AvgPool2D, Flatten
from tensorflow.keras.layers import BatchNormalization
from keras.utils import np_utils

(X_train, y_train), (X_test, y_test) = cifar100.load_data()

print("Shape of Train Set (X) :", X_train.shape)
print("Shape of Train Set (y) :", y_train.shape)
print("Shape of Test Set (X)   :", X_test.shape)
print("Shape of Test Set (y)   :", y_test.shape)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')

# normalizing the data to help with the training
X_train /= 255
X_test /= 255

Shape of Train Set (X) : (50000, 32, 32, 3)
Shape of Train Set (y) : (50000, 1)
Shape of Test Set (X)   : (10000, 32, 32, 3)
Shape of Test Set (y)   : (10000, 1)


In [2]:
n_classes = 100
print("Shape before one-hot encoding: ", y_train.shape)
Y_train = np_utils.to_categorical(y_train, n_classes)
Y_test = np_utils.to_categorical(y_test, n_classes)
print("Shape after one-hot encoding: ", Y_train.shape)

Shape before one-hot encoding:  (50000, 1)
Shape after one-hot encoding:  (50000, 100)


In [5]:
model = Sequential()
 
model.add(Conv2D(64,(5,5),padding='same',activation='relu',input_shape=(32,32,3)))
model.add(BatchNormalization())
model.add(Conv2D(64,(5,5),padding='same',activation='relu'))
model.add(MaxPool2D(pool_size=(2,2)))
model.add(Dropout(0.5))

 
model.add(Conv2D(64,(5,5),padding='same',activation='relu'))
model.add(Conv2D(64,(5,5),padding='same',activation='relu'))
model.add(AvgPool2D(pool_size=(2,2)))
model.add(Dropout(0.5))

model.add(Conv2D(64,(5,5),padding='same',activation='relu'))
model.add(Conv2D(64,(5,5),padding='same',activation='relu'))
model.add(MaxPool2D(pool_size=(2,2)))
model.add(Dropout(0.5))

model.add(Conv2D(64,(5,5),padding='same',activation='relu'))
model.add(Conv2D(64,(5,5),padding='same',activation='relu'))
model.add(AvgPool2D(pool_size=(2,2)))
model.add(Dropout(0.25))
 
model.add(Dense(16, activation='relu'))
model.add(Dropout(0.3))
model.add(Flatten())
model.add(Dense(100,activation='softmax'))
model.summary()

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_16 (Conv2D)           (None, 32, 32, 64)        4864      
_________________________________________________________________
module_wrapper_2 (ModuleWrap (None, 32, 32, 64)        256       
_________________________________________________________________
conv2d_17 (Conv2D)           (None, 32, 32, 64)        102464    
_________________________________________________________________
max_pooling2d_4 (MaxPooling2 (None, 16, 16, 64)        0         
_________________________________________________________________
dropout_10 (Dropout)         (None, 16, 16, 64)        0         
_________________________________________________________________
conv2d_18 (Conv2D)           (None, 16, 16, 64)        102464    
_________________________________________________________________
conv2d_19 (Conv2D)           (None, 16, 16, 64)       

In [6]:
model.compile(loss='categorical_crossentropy', metrics=['accuracy'], optimizer='adam')

model.fit(X_train, Y_train, batch_size=1000, epochs=50)

Epoch 1/50
50/50 [==============================] - 396s 7s/step - loss: 4.5917 - accuracy: 0.0118
Epoch 2/50
50/50 [==============================] - 384s 8s/step - loss: 4.5122 - accuracy: 0.0178
Epoch 3/50
50/50 [==============================] - 1130s 23s/step - loss: 4.4691 - accuracy: 0.0196
Epoch 4/50
50/50 [==============================] - 1297s 26s/step - loss: 4.4224 - accuracy: 0.0289
Epoch 5/50
50/50 [==============================] - 351s 7s/step - loss: 4.3392 - accuracy: 0.0362
Epoch 6/50
50/50 [==============================] - 672s 14s/step - loss: 4.1523 - accuracy: 0.0537
Epoch 7/50
50/50 [==============================] - 933s 19s/step - loss: 3.9584 - accuracy: 0.0783
Epoch 8/50
50/50 [==============================] - 951s 19s/step - loss: 3.8268 - accuracy: 0.0987
Epoch 9/50
50/50 [==============================] - 361s 7s/step - loss: 3.6830 - accuracy: 0.1214
Epoch 10/50
50/50 [==============================] - 837s 17s/step - loss: 3.5801 - accuracy: 0.1396
E

In [7]:
accuracy_train = model.evaluate(X_train, Y_train, verbose=0)[1]
accuracy_test = model.evaluate(X_test, Y_test, verbose=0)[1]

print("\nAccuracy for train set : {:.2f}".format(accuracy_train))
print("Accuracy for test set   : {:.2f}".format(accuracy_test))


Accuracy for train set : 0.49
Accuracy for test set   : 0.41


After adding convolution and dense layers, filters, padding and stride values to our model with batch size 1000 and 50 epochs we achieved 49% accuracy on train set and 41% accuracy on test set. 